### ISSUE TITLE
Implementing Neural Networks on the 10-cent RISC-V MCU without Multiplier  Tim's Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11135]
Implementing Neural Networks on the “10-cent” RISC-V MCU without Multiplier – Tim's Blog<br>
<br>
https://ift.tt/PSX7AZT<br>
<br>
<br>
<br>
<div><article>

<div>
<p>I have been meaning for a while to establish a setup to implement neural network based algorithms on smaller microcontrollers. After <a href="https://hackaday.io/project/193478-generative-ai-on-a-microcontroller/log/225316-tiny-inference-engines-for-mcu-deployment">reviewing existing solutions</a>, I felt there is no solution that I really felt comfortable with. One obvious issue is that often flexibility is traded for overhead. As always, for a really optimized solution you have to roll your own. So I did. You can find the project<a href="https://github.com/cpldcpu/BitNetMCU"> here</a> and a detailed writeup <a href="https://github.com/cpldcpu/BitNetMCU/blob/main/docs/documentation.md">here</a>. </p>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-1 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/header.png"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/header.png" data-orig-size="1425,522" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/header.png?w=300" data-attachment-id="1593" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/header.png?w=1024" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/header/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/header.png?w=840" data-image-title="header" width="1024" data-image-caption="" data-id="1593" data-comments-opened="1" class="wp-image-1593" height="375"></a></figure>
</figure>
<p>It is always easier to work with a clear challenge: I picked the CH32V003 as my target platform. This is the smallest RISC-V microcontroller on the market right now, addressing a $0.10 price point. It sports 2kb of SRAM and 16kb of flash. It is somewhat unique in implementing the RV32EC instruction set architecture, which does not even support multiplications. In other words, for many purposes this controller is less capable than an Arduino UNO.</p>
<span></span>
<p>As a test subject I chose the well-known MNIST dataset, which consists of images of hand written numbers which need to be classified from 0 to 9. Many inspiring implementation on Arduino exist for MNIST, for example <a href="https://blog.arduino.cc/2021/05/19/recognizing-handwritten-mnist-digits-on-an-arduino-uno-using-lognnet/">here</a>. In this case, the inference time was 7 seconds and 82% accuracy was achieved.</p>
<p> The idea is to train a neural network on a PC and optimize it for inference on teh CH32V003 while meetings these criteria:</p>
<ol>
<li>Be as fast and as accurate as possible</li>
<li>Low SRAM footprint during inference to fit into 2kb sram</li>
<li>Keep the weights of the neural network as small as possible</li>
<li>No multiplications!</li>
</ol>
<p>These criteria can be addressed by using a neural network with quantized weights, were each weight is represented with as few bits as possible. The best possible results are achieved when training the network already on quantized weights (Quantization Aware Training) as opposed to quantized a model that was trained with high accuracy weights. There is currently some <a href="https://arxiv.org/abs/2402.17764">hype around using Binary and Ternary weights</a> for large language models. But indeed, we can also use these approaches to fit a neural network to a small microcontroller.</p>
<p>The benefit of only using a few bits to represent each weight is that the memory footprint is low and we do not need a real multiplication instruction – inference can be reduced to additions only.</p>
<h2 class="wp-block-heading">Model structure and optimization</h2>
<p>For simplicity reasons, I decided to go for a e network architecture based on fully-connected layers instead of convolutional neural networks. The input images are reduced to a size of 16×16=256 pixels and are then fed into the network as shown below.</p>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-3 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/model_mcu.drawio.png"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/model_mcu.drawio.png" data-orig-size="808,540" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/model_mcu.drawio.png?w=300" data-attachment-id="1599" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/model_mcu.drawio.png?w=808" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/model_mcu-drawio/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/model_mcu.drawio.png?w=808" data-image-title="model_mcu.drawio" width="808" data-image-caption="" data-id="1599" data-comments-opened="1" class="wp-image-1599" height="540"></a></figure>
</figure>
<p>The implementation of the inference engine is straightforward since only fully connected layers are used. The code snippet below shows the innerloop, which implements multiplication of 4 bit weights by using adds and shifts. The weights use a one-complement encoding without zero, which helps with code efficiency. One bit, ternary, and 2 bit quantization was implemented in a similar way.</p>
<pre class="wp-block-code"><code>    int32_t sum = 0;<br>    for (uint32_t k = 0; k &lt; n_input; k+=8) {<br>        uint32_t weightChunk = *weightidx++;<p> for (uint32_t j = 0; j &lt; 8; j++) {</p>
<br>            int32_t in=*activations_idx++;<br>            int32_t tmpsum = (weightChunk &amp; 0x80000000) ? -in : in; <br>            sum += tmpsum;                                  // sign*in*1<br>            if (weightChunk &amp; 0x40000000) sum += tmpsum&lt;&lt;3; // sign*in*8<br>            if (weightChunk &amp; 0x20000000) sum += tmpsum&lt;&lt;2; // sign*in*4<br>            if (weightChunk &amp; 0x10000000) sum += tmpsum&lt;&lt;1; // sign*in*2<br>            weightChunk &lt;&lt;= 4;<br>        }<br>    }<br>    output[i] = sum;</code></pre>
<p>In addition the fc layers also normalization and ReLU operators are required. I found that it was possible to replace a more complex RMS normalization with simple shifts in the inference. Not a single full 32×32 multiplication is needed for the inference! Having this simple structure for inference means that we have to focus the effort on the training part.</p>
<p>I studied variations of the network with different numbers of bits and different sizes by varying the numer of hidden activiations. To my surprise I found that the accuracy of the prediction is proportional to the total number of bits used to store the weights. For example, when 2 bits are used for each weight, twice the numbers of weights are needed to achieve the same perforemnce as a 4 bit weight network. The plot below shows training loss vs. total number of bits. We can see that for 1-4 bits, we can basically trade more weights for less bits. This trade-off is less efficient for 8 bits and no quantization (fp32).</p>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-5 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/train_loss_vs_totalbits.png"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/train_loss_vs_totalbits.png" data-orig-size="1598,1180" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/train_loss_vs_totalbits.png?w=300" data-attachment-id="1602" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/train_loss_vs_totalbits.png?w=1024" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/train_loss_vs_totalbits/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/train_loss_vs_totalbits.png?w=840" data-image-title="train_loss_vs_totalbits" width="1024" data-image-caption="" data-id="1602" data-comments-opened="1" class="wp-image-1602" height="756"></a></figure>
</figure>
<p>I further optimized the training by using data augmentation, a cosine schedule and more epochs. It seems that 4 bit weights offered the best trade off.</p>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-7 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/explorationaugmented.png"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/explorationaugmented.png" data-orig-size="1576,1180" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/explorationaugmented.png?w=300" data-attachment-id="1604" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/explorationaugmented.png?w=1024" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/explorationaugmented/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/explorationaugmented.png?w=840" data-image-title="explorationaugmented" width="1024" data-image-caption="" data-id="1604" data-comments-opened="1" class="wp-image-1604" height="766"></a></figure>
</figure>
<p>More than 99% accuracy was achieved for 12 kbyte model size. While it is possible to achiever better accuracy with much larger models, it is significantly more accurate than other on-MCU implementations of MNIST.</p>
<h2 class="wp-block-heading">Implementation on the Microcontroller</h2>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-9 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/themcu.jpg"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/themcu.jpg" data-orig-size="1920,1612" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/themcu.jpg?w=300" data-attachment-id="1606" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/themcu.jpg?w=1024" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/themcu/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/themcu.jpg?w=840" data-image-title="themcu" width="1024" data-image-caption="" data-id="1606" data-comments-opened="1" class="wp-image-1606" height="859"></a></figure>
</figure>
<p>The model data is exported to a c-header file for inclusion into the inference code. I used the excellent <a href="https://github.com/cnlohr/ch32v003fun">ch32v003fun environment</a>, which allowed me to reduce overhead to be able to store 12kb of weights plus the inference engine in only 16kb of flash.</p>
<figure data-carousel-extra='{"blog_id":50424692,"permalink":"https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/"}' class="wp-block-gallery wp-block-gallery-11 wp-block-gallery-is-layout-flex">
<figure class="wp-block-image"><a href="https://cpldcpu.files.wordpress.com/2024/04/console.png"><img data-orig-file="https://cpldcpu.files.wordpress.com/2024/04/console.png" data-orig-size="1178,511" data-medium-file="https://cpldcpu.files.wordpress.com/2024/04/console.png?w=300" data-attachment-id="1608" data-image-meta='{"aperture":"0","credit":"","camera":"","caption":"","created_timestamp":"0","copyright":"","focal_length":"0","iso":"0","shutter_speed":"0","title":"","orientation":"0"}' data-image-description="" src="https://cpldcpu.files.wordpress.com/2024/04/console.png?w=1024" data-permalink="https://cpldcpu.wordpress.com/2024/04/24/implementing-neural-networks-on-the-10-cent-risc-v-mcu-without-multiplier/console/" data-large-file="https://cpldcpu.files.wordpress.com/2024/04/console.png?w=840" data-image-title="console" width="1024" data-image-caption="" data-id="1608" data-comments-opened="1" class="wp-image-1608" height="444"></a></figure>
</figure>
<p>There was still enough free flash to include 4 sample images. The inference output is shown above. Execution time for one inference is 13.7 ms which would actually allow to model to process moving image input in real time.</p>
<p>Alternatively, I also tested a smaller model with 4512 2-bit parameters and only 1kb of flash memory footprintg. Despite its size, it still achieves a 94.22% test accuracy and it executes in only 1.88ms.</p>
<h2 class="wp-block-heading">Conclusions</h2>
<p>This was quite a tedious projects, hunting many lost bits and rounding errors. I am quite pleased with the outcome as it shows that it is possible to compress neural networks very significantly with dedicated effort. I learned a lot and am planning to use the data pipeline for more interesting applications.</p>
</div>

</article></div><br>
<br>
<br>
<br>
<br>
<br>
via Tim's Blog https://ift.tt/Gp0s9IF<br>
<br>
April 29, 2024 at 02:47PM
